### DISCUSSION TITLE
Tracing how the code works

### DISCUSSION THREAD

**Discussion by Tobyrrr00** [yuriykulikov/AlarmClock#645] [General]
Hey @yuriykulikov 
I'm trying to add some functionality which reads text after (optional) the alarm rings. Eg. a new option 
![image](https://github.com/yuriykulikov/AlarmClock/assets/34529339/d059bf58-21f6-4f5f-8ca7-93a19f71a3de)
It would simply play another audio file (called "Blessing"), after the initial alarm is dismissed. 

Since I'm new to android dev, I'm just trying to copy the way you've been doing things. Tracing how you implement a normal alarm, and copying/editing it to suit my needs. 
For example, I created my version of com/better/alarm/alert/AlarmAlertFullScreen.kt, which doesn't have a snooze (since I don't need a snooze function for the audio reader), but also has functions for using a media player, pausing etc.

Here is my issue: I'm finding it increasingly difficult trying to trace your code, and what calls what, from what files, and in what way. Here Is what I got so far (which isn't much).
![photo_2023-12-10_20-34-25](https://github.com/yuriykulikov/AlarmClock/assets/34529339/0a1f454b-9a85-4311-b23f-74606d6923bf)

In short: this is what I'm trying to do: Alarm -> Alarm is Dismissed -> If 'blessing' option was enabled -> play blessing (below is UI for that part) -> after its done playing, dismiss everything. Very basic (for now). 
![image](https://github.com/yuriykulikov/AlarmClock/assets/34529339/ee68529c-ee98-4e11-9a15-1c54eb20aa61)

What I would like help with, is understanding how your code works. And some pointers in the direction of what I need to do, to implement those ideas.
Thank you for your time
Tobias

**Comment by yuriykulikov** [comment]
Hello Tobias,
Thank you for the interest in my project.
I have been working on some changes in the package structure which you might find helpful for understanding. Please see this branch: https://github.com/yuriykulikov/AlarmClock/tree/viewmodel 
I will probably merge this soon.

As for the implementation, the app works as following.

UI elements (fragments, activities) get data from `Store` and interact with `AlarmCore` to modify something. `AlarmCore` implements domain logic of the app, sets alarms and persists the data (using a DataStore). AlarmCore is implemented as a state machine.

When alarm fires, the event is received by a BroadcastReceiver (this is the only way) and is then given to the AlarmCore. Domain logic decides what to do with that event, most likely to start a service. That would be AlertService with it's plugins. Due to the way Android is build, the service needs to show a notification when playing the sound. This notification is a high-prio notification which is convertable to a full-screen popup (AlarmAlertFullScreen) when Android sees fit.

Now if you want to add a blessing, it would probably be easier to modify existing components to to additional stuff depending on the situation.

For example, given " If 'blessing' option was enabled " when user click on dismiss, transition to BlessingPlayingState. Send an event to the service to switch to another sound and let the AlarmAlertFullScreen know that media controls have to be rendered.
